allen-cell-animated/timelapse-colorizer
# feat: Add zoom and panning to the canvas

Problem
=======
Closes the following issues!
- closes #304
- closes #234 (kind of also closed by a previous PR, #335)
- potentially closes #285, pending check with Goutham.

(There are a few other related issues for resizing the entire viewport, and I'll leave those open for now since I think the intention is somewhat different.)

*Estimated review size: large, 50 minutes*

Solution
========
Adds the ability for the canvas to be zoomed and panned! Keyboard shortcuts and on-screen buttons are provided.

- `CanvasWrapper` can now track zoom and panning interactions, and provides buttons to do so.
  - Users can pan by holding Ctrl or Cmd and panning with the LMB, or dragging with the right or middle mouse button. 
  - Users can click the onscreen Pan toggle button to enable panning with just the LMB.
  - Users can zoom by holding Ctrl/Cmd + scrolling or by pressing the onscreen buttons. Zooming onscreen keeps the location currently under the mouse cursor fixed.
  - Panning with two fingers on the trackpad and zooming via pinch are also supported.
  - Users can reset the view with an onscreen "Reset view" button.
- Updates `ColorizeCanvas` and the two fragment shaders to update according to the new zoom and pan parameters.
- Added an icon for panning.

## Type of change
* New feature (non-breaking change which adds functionality)


Steps to Verify:
----------------
1. Open the preview link: https://allen-cell-animated.github.io/timelapse-colorizer/pr-preview/pr-347/viewer?collection=https%3A%2F%2Fdev-aics-dtp-001.int.allencell.org%2Fassay-dev%2Fcomputational%2Fdata%2Fendothileal_cell_data%2FTimelapse_20x_dataset%2Fcolorizer_outputs%2Fflow_change_movie%2Fcollections_all.json&dataset=JohnPaul_20240305_flow_change_endo_cytosol_tubulin_model_combined_cell_model&t=200
2. Hover your mouse over a cell, then zoom with the scroll wheel on mouse. Your mouse should stay in the same place over the cell even as the zoom increases.
3. Zoom in and out with the onscreen buttons.
4. Click on a cell to select it. The track should be visible and should look correct when zoomed or panned.
5. Try panning with Ctrl, Cmd (windows key on Windows), the right mouse button, the middle mouse button, and the panning button on the UI.
6. If you have a laptop, try panning and zooming with the trackpad.
7. Switch to this dataset, which has a background image enabled. The backdrop should look correct when zoomed or panned: https://allen-cell-animated.github.io/timelapse-colorizer/pr-preview/pr-347/viewer?collection=https%3A%2F%2Fdev-aics-dtp-001.int.allencell.org%2Fmicroscopy%2FClusterOutput%2FH2B_Deliverable_AnalysisPipelineOutput%2FH2B_Deliverable_InputImages_123Timelapses_Composite_Output%2Fcolorizer%2Fcollection.json&dataset=3500005807_1&bg-key=png_overlay_nucobjnum

Screenshots (optional):
-----------------------

**Video Talk-through (🔊):**

https://github.com/allen-cell-animated/timelapse-colorizer/assets/30200665/4b088939-3cbb-4050-b793-50c9b5448878


![image](https://github.com/allen-cell-animated/nucmorph-colorizer/assets/30200665/cdcdfb42-63ed-4515-849b-74bfee307d11)

![image](https://github.com/allen-cell-animated/nucmorph-colorizer/assets/30200665/aa562660-98d5-44f1-9a5a-68f7751b7598)

![image](https://github.com/allen-cell-animated/nucmorph-colorizer/assets/30200665/eb36a787-06bf-4088-ae5c-08f7f5730e9f)


Keyfiles (delete if not relevant):
-----------------------
1. `CanvasWrapper.ts`
2. `ColorizeCanvas.ts`

Thanks for contributing!
